tfhe/src/high_level_api/global_state.rs

```diff
@@ -182,16 +195,112 @@ where
 }
 
 #[cfg(feature = "gpu")]
-thread_local! {
-    static CUDA_STREAMS: std::cell::OnceCell<CudaStreams> = std::cell::OnceCell::from(CudaStreams::new_multi_gpu());
-}
+pub(in crate::high_level_api) use gpu::{
+    with_thread_local_cuda_streams, with_thread_local_cuda_streams_for_gpu_indexes,
+};
+
+#[cfg(feature = "gpu")]
+pub use gpu::CudaGpuChoice;
 
 #[cfg(feature = "gpu")]
-pub(in crate::high_level_api) fn with_thread_local_cuda_streams<
-    R,
-    F: for<'a> FnOnce(&'a CudaStreams) -> R,
->(
-    func: F,
-) -> R {
-    CUDA_STREAMS.with(|cell| func(cell.get().unwrap()))
+mod gpu {
+    use super::*;
+    use std::cell::LazyCell;
+
+    thread_local! {
+        pub(crate) static CUDA_STREAMS: RefCell<CudaStreams> = RefCell::new(CudaStreams::new_multi_gpu());
+    }
+
+    pub(in crate::high_level_api) fn with_thread_local_cuda_streams<
+        R,
+        F: for<'a> FnOnce(&'a CudaStreams) -> R,
+    >(
+        func: F,
+    ) -> R {
+        CUDA_STREAMS.with(|cell| func(&cell.borrow()))
+    }
+
+    struct CudaStreamPool {
+        multi: LazyCell<CudaStreams>,
+        single: Vec<LazyCell<CudaStreams, Box<dyn Fn() -> CudaStreams>>>,
+    }
+
+    impl CudaStreamPool {
+        fn new() -> Self {
+            Self {
+                multi: LazyCell::new(CudaStreams::new_multi_gpu),
+                single: (0..GpuIndex::num_gpus())
+                    .map(|index| {
+                        let ctor = Box::new(move || CudaStreams::new_single_gpu(GpuIndex(index)));
+                        LazyCell::new(ctor as Box<dyn Fn() -> CudaStreams>)
+                    })
+                    .collect(),
+            }
+        }
+    }
+
+    impl<'a> std::ops::Index<&'a [GpuIndex]> for CudaStreamPool {
+        type Output = CudaStreams;
+
+        fn index(&self, indexes: &'a [GpuIndex]) -> &Self::Output {
+            match indexes.len() {
+                0 => panic!("Internal error: Gpu indexes must not be empty"),
+                1 => &self.single[indexes[0].0 as usize],
+                _ => &self.multi,
+            }
+        }
+    }
+
+    impl std::ops::Index<CudaGpuChoice> for CudaStreamPool {
+        type Output = CudaStreams;
+
+        fn index(&self, choice: CudaGpuChoice) -> &Self::Output {
+            match choice {
+                CudaGpuChoice::Multi => &self.multi,
+                CudaGpuChoice::Single(index) => &self.single[index.0 as usize],
+            }
+        }
+    }
+
+    pub(in crate::high_level_api) fn with_thread_local_cuda_streams_for_gpu_indexes<
+        R,
+        F: for<'a> FnOnce(&'a CudaStreams) -> R,
+    >(
+        gpu_indexes: &[GpuIndex],
+        func: F,
+    ) -> R {
+        thread_local! {
+            static POOL: RefCell<CudaStreamPool> = RefCell::new(CudaStreamPool::new());
+        }
```

Comment:
each thread has its own static Pool of Streams ?

Reply:
Yes, This potentially could be global as is mainly to allow copying from Gpu to Cpu in case the CudaServerKey is no longer set or its not the on the same gpu, this is a bug I encountered when trying to create a CudaContext feature